File: src/nl/rubensten/texifyidea/lang/LatexCommand.java
package nl.rubensten.texifyidea.lang;

import org.jetbrains.annotations.Nullable;

/**
 * @author Ruben Schellekens
 */
public interface LatexCommand extends Dependend {

    /**
     * Looks up the given command name in all {@link LatexMathCommand}s and {@link
     * LatexNoMathCommand}.
     *
     * @param commandName
     *         The command name to look up. Can start with or without {@code \}.
     * @return The found command, or {@code null} whe the command doesn't exist.
     */
    @Nullable
    static LatexCommand lookup(String commandName) {
        if (commandName.startsWith("\\")) {
            commandName = commandName.substring(1);
        }

        LatexMathCommand math = LatexMathCommand.get(commandName);
        if (math != null) {
            return math;
        }

        return LatexNoMathCommand.get(commandName).orElse(null);
    }

    /**
     * Get the name of the command without the first backslash.
     */
    String getCommand();

    /**
     * Get the display name of the command: including backslash.
     */
    String getCommandDisplay();

    /**
     * Get the display value of the command.
     */
    String getDisplay();

    /**
     * Get all the command myArguments.
     */
    Argument[] getArguments();

    /**
     * Concatenates all arguments to each other.
     *
     * @return e.g. {@code {ARG1}{ARG2}[ARG3]...}
     */
    default String getArgumentsDisplay() {
        StringBuilder sb = new StringBuilder();
        for (Argument arg : getArguments()) {
            sb.append(arg.toString());
        }

        return sb.toString();
    }
}

File: src/nl/rubensten/texifyidea/lang/LatexCommand.kt
package nl.rubensten.texifyidea.lang

import nl.rubensten.texifyidea.psi.LatexCommands
import nl.rubensten.texifyidea.util.inMathContext
import kotlin.reflect.KClass

/**
 * @author Ruben Schellekens, Sten Wessel
 */
interface LatexCommand : Dependend {

    companion object {

        /**
         * Looks up the given command name in all [LatexMathCommand]s and [LatexNoMathCommand]s.
         *
         * @param commandName
         *          The command name to look up. Can start with or without `\`
         * @return The found command, or `null` when the command doesn't exist.
         */
        fun lookup(commandName: String?): LatexCommand? {
            var result = commandName ?: return null
            if (result.startsWith("\\")) {
                result = result.substring(1)
            }

            return LatexMathCommand[result] ?: LatexNoMathCommand[result]
        }

        /**
         * Looks up the given command within context.
         *
         * @param command The command PSI element to look up. Takes into account whether it is placed in math mode.
         * @return The found command, or `null` when the command does not exist.
         */
        fun lookup(command: LatexCommands): LatexCommand? {
            val name = command.commandToken.text
            val commandName = name.substring(1)

            return if (command.inMathContext()) {
                LatexMathCommand[commandName]
            }
            else LatexNoMathCommand[commandName]
        }
    }

    /**
     * Get the name of the command without the first backslash.
     */
    val command: String

    val commandDisplay: String
        get() = "\\$command"

    /**
     * Get the display value of the command.
     */
    val display: String?

    /**
     * Get all the command arguments.
     */
    val arguments: Array<out Argument>

    /**
     * Concatenates all arguments to each other.
     *
     * @return e.g. `{ARG1}{ARG2}[ARG3]`
     */
    fun getArgumentsDisplay(): String {
        val sb = StringBuilder()
        for (arg in arguments) {
            sb.append(arg.toString())
        }

        return sb.toString()
    }

    /**
     * Checks whether `{}` must be automatically inserted in the auto complete.
     *
     * @return `true` to insert automatically, `false` not to insert.
     */
    fun autoInsertRequired() = arguments.filter { arg -> arg is RequiredArgument }.count() >= 1

    @Suppress("UNCHECKED_CAST")
    fun <T : Argument> getArgumentsOf(clazz: KClass<T>): List<T> {
        return arguments
                .filter { clazz.java.isAssignableFrom(it.javaClass) }
                .mapNotNull { it as? T }
    }

    fun <T : Argument> getArgumentsOf(clazz: Class<T>) = getArgumentsOf(clazz.kotlin)
}

internal fun String.asRequired(type: Argument.Type = Argument.Type.NORMAL) = RequiredArgument(this, type)

internal fun String.asOptional(type: Argument.Type = Argument.Type.NORMAL) = OptionalArgument(this, type)